frappe/events_streaming/doctype/update_log/update_log.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2019, Frappe Technologies Pvt. Ltd. and contributors
+# For license information, please see license.txt
+
+from __future__ import unicode_literals
+import frappe
+from frappe.model.document import Document
+from frappe.utils.background_jobs import get_jobs
+from frappe.model import no_value_fields, table_fields
+
+class UpdateLog(Document):
+	pass
+
+def notify_consumers(doc, method=None):
+	'''Send update notification updates to event consumers whenever update log is generated'''
+	enqueued_method = 'frappe.events_streaming.doctype.event_consumer.event_consumer.notify_event_consumers'
+	jobs = get_jobs()
+	if not jobs or enqueued_method not in jobs[frappe.local.site]:
+		frappe.enqueue(enqueued_method, doctype=doc.ref_doctype, queue='long', enqueue_after_commit=True)
+
+def get_update(old, new, for_child=False):
```

Comment:
Is there a difference between this and `get_diff` in `version.py`. Can we use that one?

Reply:
no, its pretty much the same idea. i think i have told her about replacing this with `get_diff()` before.

Reply:
Yes, there are some differences:
1. `get_diff` returns the update as nested lists:

     ![list_of_list](https://user-images.githubusercontent.com/24353136/71481792-4719de00-2825-11ea-88f4-84e4957dda3b.png)

     which is not very readable when processed for updating a doc.

      `get_update` returns nested dictionaries:

       
     ![dictionary](https://user-images.githubusercontent.com/24353136/71481853-a11aa380-2825-11ea-91fa-5e8b72560ab3.png)

      which is much more readable and easier to process, especially for child table updates where updated records can be accessed using get(). Lists here are comparatively slower in terms of lookups. As these updates are going to be stored in the `Event Update Log` and `Event Sync Log` it is better if they are readable.

2. If it has nested dictionaries rather than nested lists "Update" type log will be consistent like "Create" log:
   
   ![create_type](https://user-images.githubusercontent.com/24353136/71482540-1a67c580-2829-11ea-9725-c9a26063116b.png)

3. `get_diff` also returns old values apart from the new ones which are not needed.

      ![old_new](https://user-images.githubusercontent.com/24353136/71482669-de813000-2829-11ea-9095-fd7c5af8a12b.png)

4. `get_diff` also captures `data import` if its set